app/src/authFetch.ts

```diff
@@ -0,0 +1,60 @@
+import { BASE_URL } from "@phoenix/config";
+
+const REFRESH_URL = BASE_URL + "/auth/refresh";
+
+class UnauthorizedError extends Error {
+  constructor() {
+    super("Unauthorized");
+  }
+}
+
+/**
+ * A wrapper around fetch that retries the request if the server returns a 401.
+ */
+export async function authFetch(
+  input: RequestInfo | URL,
+  init?: RequestInit
+): Promise<Response> {
+  try {
+    return await fetch(input, init).then((response) => {
+      if (response.status === 401) {
+        // If the server returns a 401, we should try to refresh the token
+        throw new UnauthorizedError();
+      }
+      return response;
+    });
+  } catch (error) {
+    if (error instanceof UnauthorizedError) {
+      // If the server returns a 401, we should try to refresh the token
+      await refreshTokens();
+      // Retry the original request
+      return fetch(input, init);
+    }
+  }
+  throw new Error("An unexpected error occurred while fetching data");
+}
+
+let refreshPromise: Promise<Response> | null = null;
+
+async function refreshTokens(url: string = REFRESH_URL): Promise<Response> {
```

Comment:
why is this not nullable yet has a fallback? also it's not passed in above can we just use the constant here everytime instead?

Reply:
Yeah we can - I was just trying to not hard code things but I guess it's kinda stupid